### ISSUE TITLE
Bug with shared weights 

### ISSUE THREAD

**Issue by anglil** [tensorflow/tensor2tensor#340]
In the T2TModel template, "If vocabularies differ, unset shared_embedding_and_softmax_weights", and this should be true for transformers. However, immediately after the unsetting of shared_embedding_and_softmax_weights, this flag is overwritten at https://github.com/tensorflow/tensor2tensor/blob/ffc580008c820d673a127e258a258a53f4a43992/tensor2tensor/utils/t2t_model.py#L103 wherein this flag is set True in transformer_base and all its inheritors. Is this conflict a bug? For language translation problems such as the walkthrough, shouldn't the weight sharing be false?

**Comment by rsepassi** [comment]
This has to do with the Problem you're running on, which defines the input and target modality and vocab size. You cannot actually share the weights if the vocab sizes differ, so the hparam is unset if that's the case.

**Comment by weitaizhang** [comment]
I think the haparam "shared_embedding_and_softmax_weights" means "share target embedding and softmax weights" and has nothing to do with the input embedding. So yes , it is a bug. "share input embedding and target embedding" should has another haparam .

**Comment by martinpopel** [comment]
By the way, [Nematus](https://github.com/EdinburghNLP/nematus#network-parameters) has two separate (and clearly documented) parameters:

| parameter            | description |
|---                   |--- |
| --tie_decoder_embeddings | tie the input embeddings of the decoder with the softmax output embeddings |
| --tie_encoder_decoder_embeddings | tie the input embeddings of the encoder and the decoder (first factor only). Source and target vocabulary size must be the same |
